lib/Entry.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Entry
+  ( dateModified
+  , Entry, EntryConvertible (..), newEntry
+  , path, masterField, fields
+  , Field (..), FieldKey (..)
+  , newField, newFieldKey, getFieldKey, newEntryTag, getEntryTag
+  , private, value, tags
+  )
+where
+
+
+import qualified Data.Text as T
+import qualified Data.HashMap.Strict as HS
+
+import Control.Lens
+import qualified Data.Aeson.Types as A
+import Data.Hashable (Hashable)
+import GHC.Generics (Generic)
+import Data.Time (UTCTime)
+import Language.Haskell.TH.Lens (_CaseE)
+
+type DateTime = UTCTime
+
+newtype FieldKey = FieldKey T.Text
```

Comment:
When a data type holds internal invariants (all characters must be valid & has at least 1 character) and is supposed to be constructed via a "smart constructor" (like `newFieldKey`), it's customary to add the prefix `Unsafe` to the constructor name.

For example, here's a quick query of the [morley codebase](https://gitlab.com/morley-framework/morley).
![image](https://user-images.githubusercontent.com/718166/154357242-3fd26442-23bf-4f20-aa14-146a64e78397.png)

This helps draw attention to the fact that usage of this constructor is discouraged and, when used, should be used with care.

```suggestion
newtype FieldKey = UnsafeFieldKey T.Text
```

Reply:
Ok, I didn't know either of those things. I'll try to tackle all of it

Reply:
I'll do the `Mk` thing later as that will affect all our other branches, let's not make this merge any more confusing and horrible

Reply:
> let's not make this merge any more confusing

Agreed. @Heimdell can you please create an issue so we won't forget to do this later?